### PR TITLE
[schemes/Typography] Add useCurrentContentSizeCategoryWhenApplied.

### PR DESCRIPTION
**Pull request by jverkoey**
This new property is a replacement for `mdc_adjustsFontForContentSizeCategory`, with the name more clearly demonstrating its intended use case.

Expected client usage:

```objc
typographyScheme = [[MDCTypographyScheme alloc] initWithDefaults:
                        MDCTypographySchemeDefaultsMaterial201902];
typographyScheme.useCurrentContentSizeCategoryWhenApplied = YES;
containerScheme.typographyScheme = typographyScheme;
button.adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable = NO;
[button applyThemeWithScheme:containerScheme];
```

Expected theming extension implementation usage (pseudo-code):

```objc
- applyThemeWithScheme:(id<MDCContainerScheming>)scheme {
 if ([scheme.typographyScheme respondsToSelector:@selector(useCurrentContentSizeCategoryWhenApplied)]
         && scheme.typographyScheme.useCurrentContentSizeCategoryWhenApplied) {
    UIContentSizeCategory sizeCategory = UIContentSizeCategoryLarge;
    if (@available(iOS 10.0, *)) {
      sizeCategory = self.traitCollection.preferredContentSizeCategory;
    } else if ([UIApplication mdc_safeSharedApplication]) {
      sizeCategory = [UIApplication mdc_safeSharedApplication].preferredContentSizeCategory;
    }
    [self setTitleFont:[scheme.typographyScheme.button mdc_scaledFontForSizeCategory:sizeCategory]
              forState:UIControlStateNormal];
  } else {
    [self setTitleFont:scheme.typographyScheme.button forState:UIControlStateNormal];
  }
}
```

Issue is being mirrored. Will update this description once it is available.

### CODE REVIEW

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 1]
```diff
@@ -90,9 +90,27 @@
...
+
+ @warning This API will eventually be deprecated. Please use
+ @c useCurrentContentSizeCategoryWhenApplied instead.
 */
 @property(nonatomic, readonly) BOOL mdc_adjustsFontForContentSizeCategory;
 
+@optional
+
+/**
+ Whether applications of this scheme to components should apply fonts in respect to the current
```
Comment (yarneo) [inline components/schemes/Typography/src/MDCTypographyScheme.h]: I didn't fully understand this sentence, the "scheme to components" part is where I got stuck grammatically

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 2]
```diff
@@ -101,6 +119,8 @@
 typedef NS_ENUM(NSInteger, MDCTypographySchemeDefaults) {
   /**
    The Material defaults, circa April 2018.
+
+   useCurrentContentSizeCategoryWhenApplied will be NO by default.
```
Comment (yarneo) [inline components/schemes/Typography/src/MDCTypographyScheme.h]: tip: Add a "Note:" prefix

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 3]
```diff
@@ -110,6 +130,8 @@ typedef NS_ENUM(NSInteger, MDCTypographySchemeDefaults) {
    This scheme implements fonts with the similar metrics as
    MDCTypographySchemeDefaultsMaterial201804 with the addition that vended fonts will have
    appropriate scalingCurves attached.
+
+   useCurrentContentSizeCategoryWhenApplied will be YES by default.
```
Comment (yarneo) [inline components/schemes/Typography/src/MDCTypographyScheme.h]: tip: Add a "Note:" prefix

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 4]
```diff
@@ -134,7 +156,19 @@ typedef NS_ENUM(NSInteger, MDCTypographySchemeDefaults) {
 @property(nonatomic, nonnull, readwrite, copy) UIFont *caption;
 @property(nonatomic, nonnull, readwrite, copy) UIFont *button;
 @property(nonatomic, nonnull, readwrite, copy) UIFont *overline;
-@property(nonatomic, readwrite) BOOL mdc_adjustsFontForContentSizeCategory;
+
+/**
+ Whether the application of this scheme to components should apply fonts in respect to the current
```
Comment (yarneo) [inline components/schemes/Typography/src/MDCTypographyScheme.h]: same here, I didn't grammatically fully understand this sentence, specifically on the "to" word connecting both parts of the sentence.

File: components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m [hunk 5]
```diff
@@ -149,4 +149,22 @@ - (void)testTypographyScheme201902ReturnsDefaultSizedFonts {
       mdc_isSimplyEqual:[scheme201902.overline mdc_scaledFontAtDefaultSize]]);
 }
 
+- (void)testTypographyScheme201804DisableUseCurrentContentSizeCategoryWhenAppliedByDefault {
```
Comment (yarneo) [inline components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m:152]: suggestion: add a test to make sure mdc_adjustsFontForContentSizeCategory is imitating the useCurrentContentSizeCategoryWhenApplied behavior.

File: components/schemes/Typography/src/MDCTypographyScheme.h (hunk 1 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MDCTypographyScheme.h]: Took another stab at it. What do you think?

File: components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m (hunk 5 above)
Comment (jverkoey) [reply components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m:152]: Good idea! Added tests.

**Review by yarneo — APPROVED** [review]

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 6]
```diff
@@ -90,9 +90,26 @@
...
+@optional
+
+/**
+ A hint for how fonts in this scheme should be applied to components in relation to Dynamic Type.
+
+ If this flag is enabled and the typography scheme's font is scalable with Dynamic Type, then fonts
+ should be adjusted for the current Dynamic Type content size category prior to being assigned to
+ the component.
+
+ @note This flag will become required in the future as a replacement for
```
Comment (codeman7) [inline components/schemes/Typography/src/MDCTypographyScheme.h:108]: _Tip:_ _"replacement for @c mdc_adjustsFontForContentSizeCategory"_

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 7]
```diff
@@ -101,6 +118,9 @@
 typedef NS_ENUM(NSInteger, MDCTypographySchemeDefaults) {
   /**
    The Material defaults, circa April 2018.
+
+   MDCTypographyScheming's useCurrentContentSizeCategoryWhenApplied is assigned to NO by default
```
Comment (codeman7) [inline components/schemes/Typography/src/MDCTypographyScheme.h:122]: _Tip:_
```diff
-  MDCTypographyScheming's useCurrentContentSizeCategoryWhenApplied is assigned to NO by default with this version.
+ @c useCurrentContentSizeCategoryWhenApplied defaults to @c NO.
```

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 8]
```diff
@@ -110,6 +130,9 @@ typedef NS_ENUM(NSInteger, MDCTypographySchemeDefaults) {
    This scheme implements fonts with the similar metrics as
    MDCTypographySchemeDefaultsMaterial201804 with the addition that vended fonts will have
    appropriate scalingCurves attached.
+
+   MDCTypographyScheming's useCurrentContentSizeCategoryWhenApplied is assigned to YES by default
```
Comment (codeman7) [inline components/schemes/Typography/src/MDCTypographyScheme.h:134]: _Tip:_
```diff
-  MDCTypographyScheming's useCurrentContentSizeCategoryWhenApplied is assigned to NO by default with this version.
+ @c useCurrentContentSizeCategoryWhenApplied defaults to @c YES.
```

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 9]
```diff
@@ -134,7 +157,20 @@ typedef NS_ENUM(NSInteger, MDCTypographySchemeDefaults) {
...
+ @note Enabling this flag only has an effect if the fonts stored on this scheme are scalable. See
+ MDCTypographySchemeDefaults for default versions that are scalable. Alternatively, you can specify
+ custom scalable fonts using the MDCFontScaler API.
+
+ When fonts are applied to components:
+
+ - If this flag is disabled, make no changes to the font.
+ - If this flag is enabled, adjust the font with respect to the current content size category.
+ */
+@property(nonatomic, readwrite) BOOL useCurrentContentSizeCategoryWhenApplied;
```
Comment (codeman7) [inline components/schemes/Typography/src/MDCTypographyScheme.h]: _Tip:_ I believe this file uses explicit property annotations, maybe beneficially to add `assign`.

File: components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m [hunk 10]
```diff
@@ -149,4 +149,47 @@ - (void)testTypographyScheme201902ReturnsDefaultSizedFonts {
...
```
Comment (codeman7) [inline components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m:152]: Should we add documentation to these method?

**Review by codeman7 — APPROVED** [review]
Looks great! I was very nit-picky but overall LGTM! I don't know if the nit-picks are worth a follow up so I'll leave that up to you.

File: components/schemes/Typography/src/MDCTypographyScheme.h (hunk 7 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MDCTypographyScheme.h:122]: I think the latter is a bit less clear if someone doesn't know what a `useCurrentContentSizeCategoryWhenApplied` property is or where it's configured.

File: components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m (hunk 10 above)
Comment (jverkoey) [reply components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m:152]: Generally-speaking unit tests shouldn't need docs if the method names are self-documenting. Do you feel these methods could use clarification?

**Comment by jverkoey** [comment]
Closes https://github.com/material-components/material-components-ios/issues/7467

**Pull request merged**